src/memtest.rs

```diff
@@ -0,0 +1,456 @@
+/// Read and compare the two halves of the memory region
+pub unsafe fn test_or(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestOr");
+    test_two_regions(base_ptr, count, timeout_checker, write_or)
+}
+
+/// Reset all bytes in specified memory region to 0xff
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// For each pair, write the AND result of a random value and the value read from the location
+/// Read and compare the two halves of the memory region
+pub unsafe fn test_and(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    debug!("Running TestAnd");
+    test_two_regions(base_ptr, count, timeout_checker, write_and)
+}
+
+/// Base function for `test_xor`, `test_sub`, `test_mul`, `test_div`, `test_or` and `test_and`
+/// Reset all bytes in specified memory region to 0xff
+/// Split specified memory region into two halves and iterate through memory locations in pairs
+/// Write to each pair using the given `write_val` function
+/// Read and compare the two halves of the memory region
+unsafe fn test_two_regions<F>(
+    base_ptr: *mut usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+    write_val: F,
+) -> Result<MemtestOutcome, MemtestError>
+where
+    F: Fn(*mut usize, *mut usize, usize),
+{
+    let half_count = count / 2;
+    let half_ptr = base_ptr.add(half_count);
+    let expected_iter =
+        u64::try_from(half_count * 2).context("Failed to convert expected_iter to u64")?;
+    timeout_checker.init(expected_iter);
+    mem_reset(base_ptr, count);
+
+    for i in 0..half_count {
+        timeout_checker.check()?;
+        write_val(base_ptr.add(i), half_ptr.add(i), random());
+    }
+
+    compare_regions(base_ptr, half_ptr, half_count, timeout_checker)
+}
+
+unsafe fn mem_reset(base_ptr: *mut usize, count: usize) {
+    write_bytes(base_ptr, 0xff, count);
+}
+
+unsafe fn compare_regions(
+    base_ptr_1: *const usize,
+    base_ptr_2: *const usize,
+    count: usize,
+    timeout_checker: &mut TimeoutChecker,
+) -> Result<MemtestOutcome, MemtestError> {
+    for i in 0..count {
+        timeout_checker.check()?;
+
+        let ptr1 = base_ptr_1.add(i);
+        let ptr2 = base_ptr_2.add(i);
+        let val1 = read_volatile(ptr1);
+        let val2 = read_volatile(ptr2);
+
+        if val1 != val2 {
+            info!("Test failed at {ptr1:?} compared to {ptr2:?}");
+            return Ok(MemtestOutcome::Fail(MemtestFailure::MismatchedValues {
+                address1: ptr1 as usize,
+                value1: val1,
+                address2: ptr2 as usize,
+                value2: val2,
+            }));
+        }
+    }
+    Ok(MemtestOutcome::Pass)
+}
+
+fn write_xor(ptr1: *mut usize, ptr2: *mut usize, val: usize) {
```

Comment:
In Rust, it's allowed to have a `fn` contained in another function. Perhaps it would be cleaner and do a better job of keeping the logic together to move this into `test_xor`?